### ISSUE TITLE
"Work-horse was terminated unexpectedly" failures are not passed through custom exception handlers

### ISSUE THREAD

**Issue by AbhinavJindalDS** [rq/rq#1620]
I have a custom exception handler which I want to run with the default exception handlers. The custom handler works fine when an exception is thrown inside the function being run by the worker, but when I (manually using kill) or the OS is the one to kill the running process (OS may kill it due to heavy memory usage), the task status shows failed in django rq and I see "Moving job to FailedJobRegistry" in the logs but my custom exception handler is not called in this case. I want my handler to be called for all cases which results in task being failed.

**Comment by BobReid** [comment]
I have a similar issues. It seems like there are several scenarios where jobs end up in the Failed Jobs Registry without going through the exception handler. 

The scenario I ran into is when jobs are abandoned and are moved as a result of the registry cleanup process. 

https://github.com/rq/rq/blob/a3fba1ca1fb16599cd775265f060697a07f5866f/rq/registry.py#L178

I am trying to implement a monitoring solution and get alerted every time something enters the Failed Job Registry but that does not seem possible at the moment. 

I am happy to contribute a fix if we can agree on the direction to take. 

I am unsure if the exception_handler is the right course of action in my scenario as there is no actual exception being caught and no exec_info to pass to the handler.

**Comment by selwin** [comment]
Yeah, this is because custom exception handlers are actually attached to the workers running jobs. When a worker is killed by the OS, it stops immediately without being able to run any exception handling tasks (including moving job to failed job registry).

The jobs that are dead will later be moved to `FailedJobRegistry` by a different worker that checks for expired jobs in the `FailedJobRegistry`. So in this case what we can do are:
1. Run expired jobs through custom exception handlers defined by the workers
2. Introduce a new `expired_job_handlers` that get executed when workers find other expired jobs and move them to the `FailedJobRegistry`.

I think option 2 is better because it's more explicit and allows you to have different handlers for different cases.

**Comment by BobReid** [comment]
Yea that makes sense.

I am OK with option 2 as well. 

If it is possible to have compatible signatures on the handlers, you could register the same handler for both cases is you do not care about the exact case, but the added flexibility is a good idea.
